pkg/webhook/elasticquota/quota_topology.go

```diff
@@ -0,0 +1,664 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package elasticquota
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	quotav1 "k8s.io/apiserver/pkg/quota/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/scheduler-plugins/pkg/apis/scheduling/v1alpha1"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	"github.com/koordinator-sh/koordinator/pkg/scheduler/plugins/elasticquota/core"
+	utilclient "github.com/koordinator-sh/koordinator/pkg/util/client"
+)
+
+type quotaTopology struct {
+	lock sync.Mutex
+	// quotaInfoMap stores all quota information
+	quotaInfoMap map[string]*core.QuotaInfo
+	// quotaHierarchy stores the quota's all children
+	quotaHierarchyInfo map[string]map[string]struct{}
+	// podCache stores the quota's all pods
+	podCache map[string]map[string]struct{}
+	client   client.Client
+}
+
+func NewQuotaTopology(client client.Client) *quotaTopology {
+	topology := &quotaTopology{
+		quotaInfoMap:       make(map[string]*core.QuotaInfo),
+		quotaHierarchyInfo: make(map[string]map[string]struct{}),
+		podCache:           make(map[string]map[string]struct{}),
+		client:             client,
+	}
+	return topology
+}
+
+func (qt *quotaTopology) OnQuotaAdd(obj interface{}) {
+	klog.V(5).Infof("OnAdd Quota to cache:%v", obj)
+
+	quota := toElasticQuota(obj)
+	if quota == nil {
+		klog.Errorf("cannot convert to *ElasticQuota:%v", obj)
+		return
+	}
+
+	quotaInfo := core.NewQuotaInfoFromQuota(quota)
+	qt.lock.Lock()
+	defer qt.lock.Unlock()
+
+	qt.quotaInfoMap[quotaInfo.Name] = quotaInfo
+	qt.podCache[quotaInfo.Name] = make(map[string]struct{})
+	if _, ok := qt.quotaHierarchyInfo[quotaInfo.Name]; !ok {
+		qt.quotaHierarchyInfo[quotaInfo.Name] = make(map[string]struct{})
+	}
+	if quotaInfo.ParentName != extension.RootQuotaName {
```

Comment:
why except rootName?

Reply:
add defaultQuotaGroup to quotaHierarchy once the quotaTopology is created like gqm do?